GenomicsDBImport incremental support for #4773

mlathara opened the pull request:
Support for incrementally adding samples to existing genomicsdb workspaces. I've added these comments to the docs, but just wanted to call out again that we recommend making a backup of the existing workspace before trying to update the workspace. Otherwise, if the incremental update fails the workspace may be in a corrupted/inconsistent state. 

If the user chooses not to backup (or can't), there is a (somewhat painful, manual) way to restore the workspace on failure IFF the --consolidate option has not been used. The tool will output a backup callset file (suffixed .inc.backup) and a file suffixed .fragmentlist with a list of all the original fragments. In order to roll back to a consistent workspace, the user must
- replace the callset file in the workspace with the one suffixed .inc.backup. That is, something like:
> mv _workspace_/callset.json.inc.backup _workspace_/callset.json
- delete all the directories in the workspace not named genomicsdb_meta_dir or included the file suffixed .fragmentlist

ldgauthier commented:
I understand that `--consolidate` will prevent your backup solution from working, but can samples still be added to a workspace that was created with `--consolidate`?

mlathara commented:
Yes, samples can be added to a workspace that was created with `--consolidate`.

And if users really want to use `--consolidate` for the incremental addition, they can backup the workspace before incremental import

mlathara commented:
@ldgauthier @droazen 

Nalini was kind enough to pull these changes into a branch so that the travis tests get run: https://github.com/broadinstitute/gatk/commits/genomicsdb_incremental_support

Let us know if you need more information from us.

ldgauthier commented:
Sorry @mlathara -- I made a branch from your fork last week, but I got caught up in other things.  I'm very excited about this new feature thought!  I was almost done setting up a large scale test and hopefully I can finish that up next week.

nalinigans commented:
@ldgauthier, @droazen, @lbergelson, just wondering if one of you review this PR?

droazen commented:
@ldgauthier said that she should be able to take a final look at this one later this week

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -263,9 +289,6 @@
     private ExecutorService inputPreloadExecutorService;
 
-    @Override
-    public boolean requiresIntervals() { return true; }
```
Do we still check for intervals when creating a new GenomicsDB?

mlathara replied:
Yes, initializeIntervals throws an exception if intervals aren't specified and we're not doing incremental import

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -564,17 +599,35 @@ private Void logMessageOnBatchCompletion(final BatchCompletionCallbackFunctionAr
         }).collect(Collectors.toList());
     }
 
+    private void generateIntervalListFromWorkspace() {
```
It would be really nice to be able to call just this method from the command line as like a dry-run mode or staging or something.

ldgauthier commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -872,6 +880,64 @@ public void testYouCantWriteIntoAnExistingDirectory(){
+        writeToGenomicsDB(LOCAL_GVCFS.subList(2,3), INTERVAL, workspace, 0, false, 0, 1, false, false, true);
+        checkJSONFilesAreWritten(workspace);
+
+        checkGenomicsDBAgainstExpected(workspace, INTERVAL, COMBINED_WITH_GENOTYPES, b38_reference_20_21, true, ATTRIBUTES_TO_IGNORE, true, false);
```
I would strongly prefer that this be a comparison to CombineGVCFs behavior or compare importing one sample and then the other to a single import with both samples.  The goal is for this functionality to be equivalent, so I'd rather show that it is than rely on expected outputs.

mlathara replied:
So, I'm actually piggybacking on an existing expected output instead of introducing a new one. As such, I think this should be checking something like the latter (import all samples at once and comparing against expected output, then break up into  multiple sets of samples that get incrementally imported and compare against the same expected output). The "import all samples at once" part is checked via this existing test:

```
@Test
public void testGenomicsDBThreeLargeSamplesWithGenotypes() throws IOException {
    ArrayList<SimpleInterval> intervals = new ArrayList<SimpleInterval>(Arrays.asList(new SimpleInterval("chr20", 1, 64444167)));
    testGenomicsDBImporterWithGenotypes(LOCAL_GVCFS, intervals, COMBINED_WITH_GENOTYPES, b38_reference_20_21, true, true, false);
}
```
And then the incremental test checks the "break up into  multiple sets of samples that get incrementally imported" part. Am I missing something?

I can certainly change it to have the incremental tests compare two workspaces instead of comparing against expected output if you prefer. I also noticed that I only have cases that test importing a single interval...will change some to multiple intervals.

ldgauthier replied:
Got it -- it wasn't obvious at first glance that the expected output was also used in another test.

ldgauthier commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -872,6 +880,64 @@ public void testYouCantWriteIntoAnExistingDirectory(){
+    }
+
+    @Test
+    public void testGenomicsDBMultipleIncrementalImports() throws IOException {
```
Same here -- instead of expected, compare with one batch of three or the CombineGVCFs output.

mlathara commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -118,13 +119,28 @@
+ *     <li>The --genomicsdb-workspace-path must point to a non-existent or empty directory, unless doing incremental import</li>
  *     <li>GenomicsDBImport uses temporary disk storage during import. The amount of temporary disk storage required can exceed the space available, especially when specifying a large number of intervals. The command line argument `--tmp-dir` can be used to specify an alternate temporary storage location with sufficient space..</li>
+ *     <li>It is recommended that users backup existing genomicsdb workspaces before adding new samples using --incremental. If the tool fails during incremental import for any reason, the workspace may be in an inconsistent/corrupted state</li>
+ *     <li>If users do not have backup workspaces available while using --incremental, another potential failsafe is available if --consolidate option was not used. In this case, if the tool fails, it will leave behind a copy of the original callset file (suffixed .inc.backup) and a list of original fragment directories (suffixed .fragmentlist - containing a list of directories within the genomic ...
```
Sure thing, wherever it makes most sense to have this is fine with me.

bhanugandham replied:
@ldgauthier I agree. We could create a short blog to explain the GenomicsDBImport incremental support feature that users are so eagerly looking forward to.

mlathara replied:
I removed that advanced option text

bhanugandham replied:
You can email that information to me and Laura. Based on the content we can figure out where that information needs to go.

mlathara commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -118,13 +119,28 @@
- *     <li>The --genomicsdb-workspace-path must point to a non-existent or empty directory.</li>
+ *     <li>The --genomicsdb-workspace-path must point to a non-existent or empty directory, unless doing incremental import</li>
  *     <li>GenomicsDBImport uses temporary disk storage during import. The amount of temporary disk storage required can exceed the space available, especially when specifying a large number of intervals. The command line argument `--tmp-dir` can be used to specify an alternate temporary storage location with sufficient space..</li>
+ *     <li>It is recommended that users backup existing genomicsdb workspaces before adding new samples using --incremental. If the tool fails during incremental import for any reason, the workspace may be in an inconsistent/corrupted state</li>
```
Hm - I don't know much about Cromwell/WDL to be honest, but the crux of the issue is that some of the workspace files get updated during incremental import. Are you saying Cromwell/WDL wouldn't use incremental import?

ldgauthier replied:
I'm saying that Cromwell/WDL wouldn't modify the original workspace -- it already makes a copy to run the task on the VM, then that modified copy will be delocalized from the VM and saved as the task output.

mlathara replied:
Gotcha - if its making a copy then Cromwell shouldn't have this issue. Did you want that stated in the doc somehow?

mlathara commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -118,13 +119,28 @@
  * <ul>
  *     <li>IMPORTANT: The -Xmx value the tool is run with should be less than the total amount of physical memory available by at least a few GB, as the native TileDB library requires additional memory on top of the Java memory. Failure to leave enough memory for the native code can result in confusing error messages!</li>
- *     <li>At least one interval must be provided</li>
+ *     <li>At least one interval must be provided, unless incrementally importing new samples in which case specified intervals are ignored in favor of intervals specified in the existing workspace</li>
```
Sure - what sort of format would you like that output in? I could add another option for, say,  `--get-intervals-from-workspace` and then just use log.info to output in the format:

```
<chr>:<start>-<stop>
```
I just have to figure out what the clean way to exit successfully at that point is...

ldgauthier replied:
The format there is a GATK-style interval list, which is fine.  In the best case scenario, I would love for this mode to output a Picard-style interval_list (with a sequence dictionary header) as a new file -- some of our interval lists are thousands of lines long, so parsing that out of the stderr seems cumbersome.

mlathara commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -118,13 +119,28 @@
+ *     <li>At least one interval must be provided, unless incrementally importing new samples in which case specified intervals are ignored in favor of intervals specified in the existing workspace</li>
  *     <li>Input GVCFs cannot contain multiple entries for a single genomic position</li>
- *     <li>The --genomicsdb-workspace-path must point to a non-existent or empty directory.</li>
+ *     <li>The --genomicsdb-workspace-path must point to a non-existent or empty directory, unless doing incremental import</li>
```
I think the workspace related special casing are the only ones that may be affected - so here and the description for workspace. On the flip side, we move some of the special casing to the code and will end up making the genomicsdb-workspace-path an optional argument.

All that said, I'm open to changing it to whatever you prefer.

mlathara commented:
Performance for single import is the same, and we have CI tests on the genomicsdb side as well. Admittedly we haven't done a lot of large scale performance testing using GenomicsDBImport.

Can you elaborate on what you mean by accuracy for large scale imports?

mlathara commented:
@ldgauthier 
My thinking on not doing the larger tests was that the native code hasn't changed for this support, so performance and functionality shouldn't see anything unexpected. Additionally, we technically do "incremental import" whenever the import is batched currently. We're just extending that same paradigm to extend beyond the case where the initial GenomicsDBImport command is used.

Of course, all of this is not to say I don't want to do the larger tests...just wondering if we could capture that in a separate issue? @droazen mentioned that there's a tentative plan for a new GATK release this week and we would like to have this feature in there, if you agree. We'll work in parallel on the performance testing you requested. Thanks!

mlathara commented:
Added some changes per Laura's review. Mainly:
- added a `--genomicsdb-update-workspace-path` that specifies the path to an existing workspace for incremental import and interval_list generation cases
- removed `--incremental` since the above made it superfluous
- added an `--output-interval-list-to-file` argument that will just generate a picard style interval_list at the location specified by the argument for an existing workspace. No import done when this is used
- changed the existing tests to use multiple intervals instead of a single interval

@ldgauthier I'm not entirely sure about the picard interval_list generation. Any chance you could help with providing some expected input/output for that so that I can add a test for it? I ran through a simple test with it but not really sure what the output should look like.

ldgauthier commented:
There's an example of a Picard interval_list here: https://software.broadinstitute.org/gatk/documentation/article?id=11009
Briefly, it's tab-separated, 1-indexed, with inclusive bounds.

ldgauthier commented:
@mlathara if the sequence dictionary in the picard format ends up getting in the way of this feature making it into the release this week, then I'm happy with GATK-style and we can deal with Picard-style in a future PR.

With regards to larger scale tests, if this is the same functionality as batch import then that's very reassuring.  It's still worth doing some larger tests, but that doesn't need to be a blocker.

mlathara commented:
Right, so I'd seen that explainer of interval_list but wasn't sure how to go about generating an interval_list to test what I'm generating. Specifically, I did a small test using some of the test resource gvcfs and generated the interval_list for those - something like this:

```
./gatk GenomicsDBImport -V ./src/test/resources/large/gvcfs/HG00096.g.vcf.gz -V ./src/test/resources/large/gvcfs/HG00268.g.vcf.gz -V ./src/test/resources/large/gvcfs/NA19625.g.vcf.gz --genomicsdb-workspace-path test -L chr20
```
And then generated the interval_list for those, like this:
```
./gatk GenomicsDBImport --genomicsdb-update-workspace-path test --output-interval-list-to-file test.interval_list
```
In this case, `test.interval_list` looks like this:
```
@HD     VN:1.6
@SQ     SN:chr20        LN:64444167
@SQ     SN:chr21        LN:46709983
chr20   1       64444167        +       .
```
Is that right? I was trying to figure out what tool output/workflow I could compare this to to make sure it was correct. I tried looking at [https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.0.0/picard_util_IntervalListTools.php](this) to try and see if there was some way to use that. But doing that didn't really seem to be the same thing....

ldgauthier commented:
@mlathara sorry, I wasn't sure what you meant.

For comparison I ran `java -jar $GATKjar IntervalListTools -I ~/workspaces/gatk/src/test/resources/large/gvcfs/HG00096.g.vcf.gz -O test.interval_list --ACTION CONCAT -I ~/workspaces/gatk/src/test/resources/large/gvcfs/HG00268.g.vcf.gz --UNIQUE true -I  /Users/gauthier/workspaces/gatk/src/test/resources/large/gvcfs/NA19625.g.vcf.gz`
which gives me

mlathara commented:
Huh - not sure there is a way to get that without querying genomicsdb itself. The interval lists that I'm constructing to support incremental import basically reconstruct the intervals the user specified on initial import, whereas the command you suggest seem to specifically combine intervals based on the data in the gvcfs. 

So, as an example for `GenomicsDBImport` the user may specify `-L chr20`. The input gvcfs (as in the example gvcfs used) may only have data between chr20 positions 17959479 and 21135661. What I reconstruct as interval for incremental import is essentially `-L chr20` while it seems you may be more interested in getting back chr20 positions 17959479 and 21135661?

I'm thinking I should probably take this out entirely for now, and we can try to address this as a future PR?

ldgauthier commented:
I'm definitely happy with the import intervals.  After I thought about it some more it really shouldn't be GDB's responsibility to figure out where there's data and having the import intervals is what I want 99% of the time.  Just recapitulating the import `-L` as an interval list is great.

mlathara commented:
Sounds good. Can I take that to mean the current functionality meets your needs?

ldgauthier commented:
The functionality as you've described it is great.  I'd still like to see another little test that outputs more than one interval.

droazen commented:
@mlathara @ldgauthier Just a head's up that if we want this branch to make it into the next release, we should get it merged no later than tomorrow morning.

ldgauthier commented:
@mlathara if you want to merge for the release and add the test later that's fine with me

mlathara commented:
I should have it momentarily...just running the tests now *fingers crossed*

mlathara commented:
OK - added two tests for interval_list generation. One with a single interval, one with multiple non-adjacent intervals. Rebased to resolve conflict as well.

mlathara commented:
@droazen @ldgauthier I'm assuming the travis failure is not caused by this? looking at some of the travis logs its complaining about "Timeout waiting for network availability" and the previous travis CI in master failed too...

ldgauthier commented:
@mlathara Travis is totally borked.  The timeout may or may not have to do with the fact that everyone is rerunning tests since they're failing sporadically.  This looks good to me, so you can merge once tests pass.  Also, if you send @droazen some notes on the new functionality to add to the release notes that would be a big help.

mlathara commented:
I will send along some notes for the release. By the way, I don't think I've enough permissions to merge the PR or rerun Travis tests....

ldgauthier commented:
I had trouble restarting my Travis tests, but luckily I can restart yours. Hopefully they pass this time.
